proxy/coco/src/request.rs

```diff
@@ -0,0 +1,336 @@
+///
+/// These transitions are pictured below:
+///
+/// ```text
+///      +----------------------------------v
+///      |                             +---------+
+///      |                   +-------->+cancelled+<------+
+///      |                   |         +----+----+       |
+///      |                   |              ^            |
+///      |                   |              |            |
+/// +----+----+       +------+--+       +---+-+      +---+---+       +------+
+/// | created +------>+requested+------>+found+----->+cloning+------>+cloned|
+/// +---------+       +------+--+       +--+--+      +---+---+       +------+
+///                          |  ^-------+  |  ^------+   |
+///                          |    failed   |   failed    |
+///                          |             v             |
+///                          |          +--+------+      |
+///                          +--------->+timed out+------+
+///                                     +---------+
+/// ```
+///
+/// The `T` type parameter represents some timestamp that is chosen by the user of the `Request`
+/// API. Note that it makes it easy to test by just choosing `()` for the timestamp.
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Request<S, T> {
+    /// The identifier of the identity on the network.
+    urn: RadUrn,
+    /// The number of attempts this request has made to complete its job.
+    attempts: Attempts,
+    /// The timestamp of the latest action to be taken on this request.
+    timestamp: T,
+    /// The state of the request, as mentioned above.
+    state: S,
+}
+
+impl<S, T> Deref for Request<S, T> {
+    type Target = S;
+
+    fn deref(&self) -> &Self::Target {
+        &self.state
+    }
+}
+
+impl<S, T> From<Request<S, T>> for Gossip {
+    fn from(request: Request<S, T>) -> Self {
+        Self {
+            urn: request.urn,
+            rev: None,
+            origin: None,
+        }
+    }
+}
+
+impl<S, T> Request<S, T> {
+    /// Get the [`RadUrn`] that this `Request` is searching for.
+    pub const fn urn(&self) -> &RadUrn {
+        &self.urn
+    }
+
+    /// Get the elapsed time between the `timestamp` provided and the current timestamp of the
+    /// `Request`.
+    ///
+    /// Note that it's expected that the `timestamp` is later than the `Request`'s timestamp.
+    pub fn elapsed(&self, timestamp: T) -> T::Output
+    where
+        T: Sub<T> + Clone,
+    {
+        timestamp - self.timestamp.clone()
+    }
+
+    /// Transition this `Request` into an `Cancelled` state. We can only transition a particular
+    /// subset of the states which are: `{Created, Requested, Found, Cloning, Cancelled}`.
+    ///
+    /// That is, attempting to cancel a `Cloned` `Request` is not permitted and will complain at
+    /// compile time.
+    pub fn cancel(self, timestamp: T) -> Request<Cancelled, T>
```

Comment:
I am not understanding why we need to pass a `timestamp` when we cancel a request. Is it so to capture when the request was cancelled?

Reply:
You actually understood correctly :) The timestamp is to record the last action. Which plays nicely with `filter` for getting items that may have elapsed some `delta`